higher/index.ts

```diff
@@ -0,0 +1,2314 @@
+
+export type payloadType = {
+  msgType: string,
+  // when msgType = REQ_UPDATE_LINKED
+  tempName?: string,
+  srcIdkey?: string,
+  newLinkedMembers?: groupObjType[],
+  // when msgType = CONFIRM_UPDATE_LINKED
+  existingGroups?: storageObjType[],
+  existingData?: storageObjType[],
+  // when msgType = REQ_CONTACT
+  reqIdkey?: string,
+  reqContactName?: string,
+  reqContactGroups?: groupObjType[],
+  // when msgType = CONFIRM_CONTACT
+  contactName?: string,
+  contactGroups?: groupObjType[],
+  // when msgType = UPDATE_DATA
+  key?: string,
+  dataValue?: any,
+  // when msgType = UPDATE_GROUP, ADD_ADMIN, etc
+  groupID?: string,
+  // when msgType = UPDATE_GROUP
+  groupValue?: groupValType,
+  // when msgType = LINK_GROUPS, ADD/REMOVE_PARENT/CHILD
+  parentID?: string,
+  childID?: string,
+  // when msgType = ADD/REMOVE_ADMIN
+  adminID?: string,
+  // when msgType = ADD/REMOVE_WRITER
+  writerID?: string,
+  // when msgType = 
+  // when msgType = 
+  // when msgType = 
+  // when msgType = 
+  // when msgType = 
+  // when msgType = 
+  // FIXME many variable fields
+};
+
+type storageObjType = {
+  key: string,
+  value: any // FIXME groupObjType | dataObjType
+};
+
+export class Higher {
+  // TODO make variables private
+  static SLASH   : string = "/";
+  static DATA    : string = "__data";
+  static GROUP   : string = "__group";
+  static LINKED  : string = "__linked";
+  static CONTACTS: string = "__contacts";
+  static OUTSTANDING_IDKEY: string = "__outstandingIdkey";
+
+  // valid message types
+  static REQ_UPDATE_LINKED    : string = "requestUpdateLinked";
+  static CONFIRM_UPDATE_LINKED: string = "confirmUpdateLinked";
+  static REQ_CONTACT          : string = "requestContact";
+  static CONFIRM_CONTACT      : string = "confirmContact";
+  static LINK_GROUPS          : string = "linkGroups";
+  static ADD_PARENT           : string = "addParent";
+  static ADD_CHILD            : string = "addChild";
+  static ADD_WRITER           : string = "addWriter";
+  static ADD_ADMIN            : string = "addAdmin";
+  static REMOVE_PARENT        : string = "removeParent";
+  static REMOVE_WRITER        : string = "removeWriter";
+  static REMOVE_ADMIN         : string = "removeAdmin";
+  static UPDATE_GROUP         : string = "updateGroup";
+  static UPDATE_DATA          : string = "updateData";
+  static DELETE_DEVICE        : string = "deleteDevice";
+  static DELETE_GROUP         : string = "deleteGroup";
+  static DELETE_DATA          : string = "deleteData";
+
+  // default auth/unauth functions do nothing
+  defaultOnAuth  : () => void = () => {};
+  defaultOnUnauth: () => void = () => {};
+  // default callback
+  defaultValidateCallback = (payload) => {
+    console.log("validating payload... " + JSON.stringify(payload));
+    return true;
+  };
+
+
+  storagePrefixes: string[] = [Higher.GROUP];
+  onAuth  : () => void;
+  onUnauth: () => void;
+  turnEncryptionOff: boolean;
+  validateCallback: (payloadType) => boolean;
```

Comment:
To make things simpler, maybe instead of explicit "validation" or "conflict" defs in the Higher lib, this can be links to other "modules"/"libs" that perform validation, which would take a lot of logic our of the "higher" level lib and could be easier to integrate/remove by developers.